geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/AvroIngestProcessor.scala

```diff
@@ -143,41 +99,158 @@ trait AvroIngestProcessor extends FeatureTypeProcessor {
+                } catch {
+                  case NonFatal(e) =>
+                    failure += 1L
+                    logError(sf, e)
                 }
-                writer.apply(sf)
-                success += 1L
-              } catch {
-                case NonFatal(e) =>
-                  failure += 1L
-                  logError(sf, e)
               }
+            } finally {
+              writers.returnWriter(writer)
             }
           } finally {
             reader.close()
           }
         }
       })
 
-      (success, failure)
+      IngestResult(success, failure)
     }
 
-    def buildConverter(input: SimpleFeatureType, output: SimpleFeatureType): Option[SimpleFeature => SimpleFeature] = {
-      FeatureTypeProcessor.checkCompatibleSchema(input, output).map { error =>
-        if (matchMode == LenientMatch) {
-          convert(input, output)
-        } else {
-          throw error
-        }
+    private def checkSchemaAndMapping(sft: SimpleFeatureType): Option[SimpleFeature => SimpleFeature] = {
+      val existing = Try(store.getSchema(sft.getTypeName)).getOrElse(null)
+
+      if (existing == null) {
+        logger.info(s"Creating schema ${sft.getTypeName}: ${SimpleFeatureTypes.encodeType(sft)}")
+        store.createSchema(sft)
+        None
+      } else if (SimpleFeatureTypes.compare(sft, existing) == 0) {
+        None
+      } else if (mode == CompatibilityMode.Update) {
+        logger.info(
+          s"Updating schema ${sft.getTypeName}:" +
+              s"\n  from: ${SimpleFeatureTypes.encodeType(existing)}" +
+              s"\n  to:   ${SimpleFeatureTypes.encodeType(sft)}")
+        store.updateSchema(sft.getTypeName, sft)
+        None
+      } else if (mode == CompatibilityMode.Existing) {
+        Some(AvroIngestProcessor.convert(sft, existing))
+      } else {
+        throw new IllegalArgumentException(
+          "Input schema does not match existing type:" +
+              s"\n  Input:    ${SimpleFeatureTypes.encodeType(sft)}" +
+              s"\n  Existing: ${SimpleFeatureTypes.encodeType(existing)}")
       }
     }
   }
 }
 
+object AvroIngestProcessor {
+
+  import scala.collection.JavaConverters._
+
+  // noinspection ScalaDeprecation
+  private val ModeMappings =
+    Seq(Properties.ExactMatch -> CompatibilityMode.Exact, Properties.LenientMatch -> CompatibilityMode.Existing)
+
+  // noinspection ScalaDeprecation
+  object Properties {
+
+    val UseProvidedFid: PropertyDescriptor =
+      new PropertyDescriptor.Builder()
+          .name("Use provided feature ID")
+          .description("Use the feature ID from the Avro record, or generate a new one")
+          .required(false)
+          .defaultValue("false")
+          .allowableValues("true", "false")
+          .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+          .build()
+
+    @deprecated
+    val ExactMatch = "by attribute number and order"
+    @deprecated
+    val LenientMatch = "by attribute name"
```

Comment:
Is there a replacement for this?  Or an alternative approach?

Reply:
the functionality is the same, i just standardized on the 'schema compatibility' property that was in use by the other processors. so now you'd use 'exact' and 'existing' instead of these two.

Reply:
Ok, that likely needs a quick note in the upgrade guide.